File: common/cazzar/mods/jukeboxreloaded/gui/GUIJukeBox.java
package cazzar.mods.jukeboxreloaded.gui;

import static cazzar.mods.jukeboxreloaded.lib.Reference.GUIJUKEBOX_TEXTURE_FILE;
import static cazzar.mods.jukeboxreloaded.lib.Reference.GUIJukeBoxActions.NEXT;
import static cazzar.mods.jukeboxreloaded.lib.Reference.GUIJukeBoxActions.PLAY;
import static cazzar.mods.jukeboxreloaded.lib.Reference.GUIJukeBoxActions.PREVIOUS;
import static cazzar.mods.jukeboxreloaded.lib.Reference.GUIJukeBoxActions.REPEAT_ALL;
import static cazzar.mods.jukeboxreloaded.lib.Reference.GUIJukeBoxActions.REPEAT_OFF;
import static cazzar.mods.jukeboxreloaded.lib.Reference.GUIJukeBoxActions.REPEAT_ONE;
import static cazzar.mods.jukeboxreloaded.lib.Reference.GUIJukeBoxActions.SHUFFLE;
import static cazzar.mods.jukeboxreloaded.lib.Reference.GUIJukeBoxActions.SHUFFLE_OFF;
import static cazzar.mods.jukeboxreloaded.lib.Reference.GUIJukeBoxActions.STOP;
import static java.lang.Math.floor;

import java.util.Random;

import net.minecraft.client.gui.GuiButton;
import net.minecraft.client.gui.inventory.GuiContainer;
import net.minecraft.entity.player.EntityPlayer;

import org.lwjgl.opengl.GL11;

import cazzar.mods.jukeboxreloaded.blocks.TileJukeBox;
import cazzar.mods.jukeboxreloaded.lib.Strings;
import cazzar.mods.jukeboxreloaded.network.packets.PacketJukeboxDescription;
import cpw.mods.fml.common.network.PacketDispatcher;
import cpw.mods.fml.relauncher.Side;
import cpw.mods.fml.relauncher.SideOnly;

@SideOnly(Side.CLIENT)
public class GUIJukeBox extends GuiContainer {
	TileJukeBox	tileJukeBox;
	private TexturedButton	btnPlay, btnStop, btnShuffle, btnShuffleOff,
			btnRepeatAll, btnRepeatOne, btnRepeatOff;
	
	public GUIJukeBox(EntityPlayer player, TileJukeBox tileJukeBox) {
		super(new ContainerJukeBox(player.inventory, tileJukeBox));
		this.tileJukeBox = tileJukeBox;
		xSize = 176;
		ySize = 176;
	}
	
	@Override
	protected void actionPerformed(GuiButton btn) {
		final boolean wasPlaying = tileJukeBox.isPlayingRecord();
		
		switch (btn.id) {
			case PLAY:
				tileJukeBox.playSelectedRecord();
				break;
			case STOP:
				tileJukeBox.stopPlayingRecord();
				break;
			case NEXT:
				if (wasPlaying) tileJukeBox.stopPlayingRecord();
				if (tileJukeBox.shuffleEnabled()) {
					final Random random = new Random();
					if (tileJukeBox.getLastSlotWithItem() <= 0) return;
					final int nextDisk = random.nextInt(tileJukeBox
							.getLastSlotWithItem());
					if (tileJukeBox.getCurrentRecordNumer() != nextDisk)
						tileJukeBox.setRecordPlaying(nextDisk);
				}
				tileJukeBox.nextRecord();
				if (wasPlaying) tileJukeBox.playSelectedRecord();
				break;
			case PREVIOUS:
				if (wasPlaying) tileJukeBox.stopPlayingRecord();
				tileJukeBox.previousRecord();
				if (wasPlaying) tileJukeBox.playSelectedRecord();
				break;
			case REPEAT_ALL:
				tileJukeBox.setRepeatMode(1);
				break;
			case REPEAT_ONE:
				tileJukeBox.setRepeatMode(2);
				break;
			case REPEAT_OFF:
				tileJukeBox.setRepeatMode(0);
				break;
			case SHUFFLE:
				tileJukeBox.setShuffle(true);
				break;
			case SHUFFLE_OFF:
				tileJukeBox.setShuffle(false);
				break;
		}
		btnPlay.enabled = !(btnStop.enabled = tileJukeBox.isPlayingRecord());
		btnShuffle.enabled = !tileJukeBox.shuffleEnabled();
		btnShuffleOff.enabled = tileJukeBox.shuffleEnabled();
		switch (tileJukeBox.getReplayMode()) {
			case 0:
				btnRepeatOff.enabled = false;
				btnRepeatOne.enabled = true;
				btnRepeatAll.enabled = true;
				break;
			case 1:
				btnRepeatOff.enabled = true;
				btnRepeatOne.enabled = true;
				btnRepeatAll.enabled = false;
				break;
			case 2:
				btnRepeatOff.enabled = true;
				btnRepeatOne.enabled = false;
				btnRepeatAll.enabled = true;
				break;
		}
		
		PacketDispatcher.sendPacketToServer((new PacketJukeboxDescription(
				tileJukeBox)).makePacket());
	}
	
	@Override
	protected void drawGuiContainerBackgroundLayer(float var1, int var2,
			int var3) {
		
		
		btnPlay.enabled = !(btnStop.enabled = tileJukeBox.isPlayingRecord());
		btnShuffle.enabled = !tileJukeBox.shuffleEnabled();
		btnShuffleOff.enabled = tileJukeBox.shuffleEnabled();
		
		switch (tileJukeBox.getReplayMode()) {
			case 0:
				btnRepeatOff.enabled = false;
				btnRepeatOne.enabled = true;
				btnRepeatAll.enabled = true;
				break;
			case 1:
				btnRepeatOff.enabled = true;
				btnRepeatOne.enabled = true;
				btnRepeatAll.enabled = false;
				break;
			case 2:
				btnRepeatOff.enabled = true;
				btnRepeatOne.enabled = false;
				btnRepeatAll.enabled = true;
				break;
		}
		
		GL11.glColor4f(1.0F, 1.0F, 1.0F, 1.0F);
		mc.renderEngine.bindTexture(GUIJUKEBOX_TEXTURE_FILE);
		final int xStart = (width - xSize) / 2;
		final int yStart = (height - ySize) / 2;
		drawTexturedModalRect(xStart, yStart, 0, 0, xSize, ySize);
	}
	
	@Override
	protected void drawGuiContainerForegroundLayer(int x, int y) {
		
		final String containerName = Strings.GUI_JUKEBOX_NAME.toString();
		fontRenderer.drawString(containerName,
				xSize / 2 - fontRenderer.getStringWidth(containerName) / 2, 6,
				4210752);
		fontRenderer.drawString(
				Strings.GUI_INVENTORY.toString(), 8,
				ySize - 93, 4210752);
		
		GL11.glColor4f(1.0F, 1.0F, 1.0F, 1.0F);
		mc.renderEngine.bindTexture(GUIJUKEBOX_TEXTURE_FILE);
		
		final int xOffset = 53;
		final int yOffset = 16;
		final int size = 18;
		
		final int index = tileJukeBox.getCurrentRecordNumer();
		final int column = (int) floor(index / 4D);
		final int row = index % 4;
		
		// System.out.println(column + ":" + row + ":" + index);
		// Args: x, y, xOffset, yOffset, Width, Height
		// xOffset + (size * row), yOffset + (size * column),
		drawTexturedModalRect(0 + xOffset + size * row, 0 + yOffset + size
				* column, 176, 0, 18, 18);
	}
	
	@SuppressWarnings("unchecked")
	@Override
	public void initGui() {
		super.initGui();
		final int xStart = (width - xSize) / 2;
		final int yStart = (height - ySize) / 2;
		
		buttonList.add(btnPlay = new TexturedButton(PLAY, xStart + 7,
				yStart + 17, 20, 20, GUIJUKEBOX_TEXTURE_FILE, 176, 38, 176, 18,
				176, 58));
		buttonList.add(btnStop = new TexturedButton(STOP, xStart + 29,
				yStart + 17, 20, 20, GUIJUKEBOX_TEXTURE_FILE, 176, 98, 176, 78,
				176, 118));
		
		buttonList.add(new TexturedButton(NEXT, xStart + 29, yStart + 39, 20,
				20, GUIJUKEBOX_TEXTURE_FILE, 216, 38, 216, 18, 216, 58));
		buttonList.add(new TexturedButton(PREVIOUS, xStart + 7, yStart + 39,
				20, 20, GUIJUKEBOX_TEXTURE_FILE, 236, 38, 236, 18, 236, 58));
		
		buttonList.add(btnRepeatOne = new TexturedButton(REPEAT_ONE,
				xStart + 150, yStart + 17, 20, 20, GUIJUKEBOX_TEXTURE_FILE,
				196, 98, 196, 78, 196, 118));
		buttonList.add(btnRepeatAll = new TexturedButton(REPEAT_ALL,
				xStart + 150, yStart + 40, 20, 20, GUIJUKEBOX_TEXTURE_FILE,
				216, 98, 216, 78, 216, 118));
		buttonList.add(btnRepeatOff = new TexturedButton(REPEAT_OFF,
				xStart + 150, yStart + 63, 20, 20, GUIJUKEBOX_TEXTURE_FILE,
				196, 158, 196, 138, 196, 178));
		
		buttonList.add(btnShuffle = new TexturedButton(SHUFFLE, xStart + 128,
				yStart + 17, 20, 20, GUIJUKEBOX_TEXTURE_FILE, 236, 98, 236, 78,
				236, 118));
		buttonList.add(btnShuffleOff = new TexturedButton(SHUFFLE_OFF,
				xStart + 128, yStart + 40, 20, 20, GUIJUKEBOX_TEXTURE_FILE,
				176, 158, 176, 138, 176, 178));
	}
}

File: src/main/scala/net/cazzar/mods/jukeboxreloaded/client/GuiJukebox.scala
package net.cazzar.mods.jukeboxreloaded.client

import net.cazzar.mods.jukeboxreloaded.blocks.tileentity.TileJukebox
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.client.gui.inventory.GuiContainer
import net.cazzar.mods.jukeboxreloaded.common.gui.ContainerJukebox
import net.minecraft.client.gui.GuiButton
import org.lwjgl.opengl.GL11._
import net.cazzar.mods.jukeboxreloaded.common.{Strings, Reference}
import net.cazzar.mods.jukeboxreloaded.common.ReplayMode._
import Reference.JUKEBOX_GUI_TEXTURE
import Reference.GuiActions._
import net.cazzar.corelib.client.gui.TexturedButton
import net.cazzar.mods.jukeboxreloaded.JukeboxReloaded
import cpw.mods.fml.relauncher.Side
import cpw.mods.fml.common.network.FMLOutboundHandler
import cpw.mods.fml.common.network.FMLOutboundHandler.OutboundTarget
import net.cazzar.mods.jukeboxreloaded.networking.packets.PacketJukeboxGuiAction
import net.cazzar.corelib.lib.SoundSystemHelper

class GuiJukebox(player: EntityPlayer, tile: TileJukebox) extends GuiContainer(new ContainerJukebox(player.inventory, tile)) {
    xSize = 176
    ySize = 176

    //(width - xSize) / 2
    def xStart = (this.width - this.xSize) / 2

    //(height - ySize) / 2
    def yStart = (this.height - this.ySize) / 2

    var btnPlay: TexturedButton = null
    var btnStop: TexturedButton = null
    var btnNext: TexturedButton = null
    var btnPrev: TexturedButton = null
    var btnRepeatOne: TexturedButton = null
    var btnRepeatAll: TexturedButton = null
    var btnRepeatOff: TexturedButton = null
    var btnShuffleOn: TexturedButton = null
    var btnShuffleOff: TexturedButton = null

    override def drawGuiContainerForegroundLayer(x: Int, y: Int) = {
        val name = Strings.GUI_JUKEBOX_NAME
        fontRendererObj.drawString(name, xSize / 2 - (fontRendererObj.getStringWidth(name) / 2), 6, 4210752)
        fontRendererObj.drawString(Strings.GUI_INVENTORY, 8, ySize - 93, 4210752)

        glColor4f(1.0F, 1.0F, 1.0F, 1.0F)
        mc.renderEngine.bindTexture(JUKEBOX_GUI_TEXTURE)
        val xOffset = 53
        val yOffset = 16
        val size = 18

        val index = tile.current
        val column = Math.floor(index / 4D).asInstanceOf[Int]
        val row = index % 4

        drawTexturedModalRect(xOffset + size * row, yOffset + size * column, 176, 0, 18, 18)

        import scala.collection.JavaConversions._
        for (button <- buttonList.asInstanceOf[java.util.List[GuiButton]]) {
            button match {
                case btn: TexturedButton =>
                    if ((x >= btn.xPosition && x <= btn.xPosition + btn.getHeight) && (y >= btn.yPosition && y <= btn.yPosition + btn.getWidth))
                        if (!btn.getTooltip.trim().isEmpty && btn.enabled) {
                            btn.drawToolTip(x - xStart, y - yStart)
                        }
                case _ =>
            }
        }
    }

    def drawGuiContainerBackgroundLayer(p1: Float, p2: Int, p3: Int) = {
        updateButtonStates()
        glColor4f(1F, 1F, 1F, 1F)
        mc.renderEngine.bindTexture(Reference.JUKEBOX_GUI_TEXTURE)
        //        val xStart = (this.field_146294_l - this.field_146999_f) / 2
        //        val yStart = (this.field_146295_m - this.field_147000_g) / 2

        drawTexturedModalRect(xStart, yStart, 0, 0, xSize, ySize)
    }

    //actionPerformed
    override def actionPerformed(button: GuiButton) = {
        //this is only server -> client so we have to hack backwards.
        //tile.markForUpdate()

        JukeboxReloaded.proxy.channel.get(Side.CLIENT).attr(FMLOutboundHandler.FML_MESSAGETARGET).set(OutboundTarget.TOSERVER)
        JukeboxReloaded.proxy.channel.get(Side.CLIENT).writeOutbound(new PacketJukeboxGuiAction(tile, button.id))
    }

    def initButtons() = {
        btnPlay = new TexturedButton(this, PLAY, xStart + 7, yStart + 17, 20, 20, JUKEBOX_GUI_TEXTURE, 176, 38, 176, 18, 176, 58)
        btnStop = new TexturedButton(this, STOP, xStart + 29, yStart + 17, 20, 20, JUKEBOX_GUI_TEXTURE, 176, 98, 176, 78, 176, 118)
        btnNext = new TexturedButton(this, NEXT, xStart + 29, yStart + 39, 20, 20, JUKEBOX_GUI_TEXTURE, 216, 38, 216, 18, 216, 58)
        btnPrev = new TexturedButton(this, PREVIOUS, xStart + 7, yStart + 39, 20, 20, JUKEBOX_GUI_TEXTURE, 236, 38, 236, 18, 236, 58)
        btnRepeatOne = new TexturedButton(this, REPEAT_ONE, xStart + 150, yStart + 17, 20, 20, JUKEBOX_GUI_TEXTURE, 196, 98, 196, 78, 196, 118)
        btnRepeatAll = new TexturedButton(this, REPEAT_ALL, xStart + 150, yStart + 40, 20, 20, JUKEBOX_GUI_TEXTURE, 216, 98, 216, 78, 216, 118)
        btnRepeatOff = new TexturedButton(this, REPEAT_OFF, xStart + 150, yStart + 63, 20, 20, JUKEBOX_GUI_TEXTURE, 196, 158, 196, 138, 196, 178)
        btnShuffleOn = new TexturedButton(this, SHUFFLE, xStart + 128, yStart + 17, 20, 20, JUKEBOX_GUI_TEXTURE, 236, 98, 236, 78, 236, 118)
        btnShuffleOff = new TexturedButton(this, SHUFFLE_OFF, xStart + 128, yStart + 40, 20, 20, JUKEBOX_GUI_TEXTURE, 176, 158, 176, 138, 176, 178)

        val list: java.util.List[GuiButton] = buttonList.asInstanceOf[java.util.List[GuiButton]]
        list.add(btnPlay)
        list.add(btnStop)
        list.add(btnNext)
        list.add(btnPrev)
        list.add(btnRepeatOne)
        list.add(btnRepeatAll)
        list.add(btnRepeatOff)
        list.add(btnShuffleOn)
        list.add(btnShuffleOff)
    }

    override def initGui() = {
        super.initGui()

        initButtons()
        initTooltips()
        updateButtonStates()
    }

    def initTooltips() = {
        btnPlay.setTooltip(Strings.TOOLTIP_PLAY)
        btnStop.setTooltip(Strings.TOOLTIP_STOP)
        btnNext.setTooltip(Strings.TOOLTIP_NEXT)
        btnPrev.setTooltip(Strings.TOOLTIP_PREV)
        btnRepeatAll.setTooltip(Strings.TOOLTIP_REPEAT_ALL)
        btnRepeatOff.setTooltip(Strings.TOOLTIP_REPEAT_NONE)
        btnRepeatOne.setTooltip(Strings.TOOLTIP_REPEAT_ONE)
        btnShuffleOn.setTooltip(Strings.TOOLTIP_SHUFFLE_ON)
        btnShuffleOff.setTooltip(Strings.TOOLTIP_SHUFFLE_OFF)
    }


    def updateButtonStates() = {
//        println(SoundSystemHelper.isPlaying(mc.renderGlobal, tile.identifier) + " : " + tile.identifier.toString)
//
//        println(SoundSystemHelper.getSoundHandler.isSoundPlaying(SoundSystemHelper.getSoundForChunkCoordinates(mc.renderGlobal, tile.identifier)))
        println(SoundSystemHelper.getSoundForChunkCoordinates(mc.renderGlobal, tile.identifier).hashCode())
        btnStop.enabled = tile.playing
        btnPlay.enabled = !tile.playing
        btnShuffleOn.enabled = !tile.shuffle
        btnShuffleOff.enabled = tile.shuffle

        tile.replayMode match {
            case ALL =>
                btnRepeatAll.enabled = false
                btnRepeatOne.enabled = true
                btnRepeatOff.enabled = true
            case ONE =>
                btnRepeatAll.enabled = true
                btnRepeatOne.enabled = false
                btnRepeatOff.enabled = true
            case OFF =>
                btnRepeatAll.enabled = true
                btnRepeatOne.enabled = true
                btnRepeatOff.enabled = false
        }
    }
}
